rllib/utils/exploration/random_encoder.py

```diff
@@ -0,0 +1,338 @@
+from gym.spaces import Discrete, Space
+import numpy as np
+from typing import List, Optional, Tuple, Union
+
+from ray.rllib.models.action_dist import ActionDistribution
+from ray.rllib.models.catalog import ModelCatalog
+from ray.rllib.models.modelv2 import ModelV2
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.exploration.exploration import Exploration
+from ray.rllib.utils.framework import try_import_tf
+from ray.rllib.utils.from_config import from_config
+from ray.rllib.utils.tf_ops import get_placeholder
+from ray.rllib.utils.typing import FromConfigSpec, ModelConfigDict, TensorType
+
+tf1, tf, tfv = try_import_tf()
+
+
+class MovingMeanStd:
+    """Track moving mean, std and count."""
+
+    def __init__(self, epsilon: float = 1e-4, shape: List[int] = []):
+        """Initialize object.
+
+        Args:
+            epsilon (float): Initial count.
+            shape (List[int]): Shape of the trackables mean and std.
+        """
+        self.mean = np.zeros(shape, dtype=np.float32)
+        self.var = np.ones(shape, dtype=np.float32)
+        self.count = epsilon
+
+    def __call__(self, inputs: np.ndarray) -> np.ndarray:
+        """Normalize input batch using moving mean and std.
+
+        Args:
+            inputs (np.ndarray): Input batch to normalize.
+
+        Returns:
+            Logarithmic scaled normalized output.
+        """
+        batch_mean = np.mean(inputs, axis=0)
+        batch_var = np.var(inputs, axis=0)
+        batch_count = inputs.shape[0]
+        self.update_params(batch_mean, batch_var, batch_count)
+        return np.log(inputs / self.std + 1)
+
+    def update_params(self, batch_mean: float, batch_var: float,
+                      batch_count: int) -> None:
+        """Update moving mean, std and count.
+
+        Args:
+            batch_mean (float): Input batch mean.
+            batch_var (float): Input batch variance.
+            batch_count (float): Number of cases in the batch.
+        """
+        delta = batch_mean - self.mean
+        tot_count = self.count + batch_count
+
+        self.mean = self.mean + delta + batch_count / tot_count
```

Comment:
hmm sorry for the stupid question, but I don't quite get the formula for the moving mean here. as written, self.mean basically is:

self.mean + batch_mean - self.mean + batch_count / tot_count
which is:
batch_mean + batch_count / tot_count

why are we taking batch_mean plus a small ratio as the moving mean?

Reply:
Yes, this construct is adopted author's reference code. I have also experimented using the correct moving mean (`(self.mean * self.count + batch_mean * batch_count) / tot_count`) version, results are almost similar.

Reply:
ah ok. it's not uncommon to see questionable reference implementation.
we can keep it as is for now.